### PR TITLE
IoUring: add base implementation for IoUringWorker and IoUringSocket

### PR DESCRIPTION
**Pull request by soulxu**
Commit Message: IoUring: add base implementation for IoUringWorker and IoUringSocket
Additional Description:
This PR adds a basic interface for IoUringWorker and IoUringSocket, also provides a basic implementation.

The IoUringWorker is a per-thread worker for iouring. It owns and manages all the IoUringSockets on the thread.
The IoUringSocket is the abstract of network socket. It is used to receive/send data and receive events on a socket.

Risk Level: low
Testing: unittest and integration test.
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a
Part of issues https://github.com/envoyproxy/envoy/issues/28395

Signed-off-by: He Jie Xu [hejie.xu@intel.com](mailto:hejie.xu@intel.com)
Signed-off-by: Zhi Hao Xie [zhihao.xie@intel.com](mailto:zhihao.xie@intel.com)
Signed-off-by: Dmitry Rozhkov [dmitry.rozhkov@intel.com](mailto:dmitry.rozhkov@intel.com)

### CODE REVIEW

**Comment by soulxu** [comment]
wait https://github.com/envoyproxy/envoy/pull/28579 merge first.

**Comment by soulxu** [comment]
@KBaichoo @yanavlasov this is ready for review, thanks in advance!

File: envoy/common/io/io_uring.h
Comment (zhxie) [inline envoy/common/io/io_uring.h:162]: Actually, `IoUringWorker` in this PR does not require `IoUringSocket`. So we can just move the `IoUringWorker` definition below to here to avoid the forward declaration.
But in the future, we shall add methods with params including `IoUringSocket` to `IoUringWorker`.

File: source/common/io/io_uring_worker_impl.h
Comment (zhxie) [inline source/common/io/io_uring_worker_impl.h:11]: And I think the derived class `BaseRequest` is a little bit redundancy. Can we just merge `BaseRequest` and its abstract class `Request`? Of course it is Ok to keep the status quo.

File: test/mocks/io/BUILD
Comment (zhxie) [inline test/mocks/io/BUILD:13]: Redundant empty array `srcs`.

**Review by zhxie — APPROVED** [review]
LGTM.

`IoUringSocket` holds the exact network fd and provides an interface for IO handles to operate . `IoUringWorker` represents the event of the inner `IoUring`. After the PR getting merged, we can then start implementing IO handle of io_uring which can be used in all Envoy connections.

File: source/common/io/io_uring_worker_impl.h
Comment (soulxu) [reply source/common/io/io_uring_worker_impl.h:11]: yea, good point, it could be merged, let me try.

File: envoy/common/io/io_uring.h
Comment (soulxu) [reply envoy/common/io/io_uring.h:162]: Yea, as you said, very soon we will add a set of submit*Request method https://github.com/envoyproxy/envoy/pull/24958/files#diff-dad289fe5a60a8403503d6ec8e79baeb72f4de7a40aaba36646494cdfde1fd39R402, then the IoUringWorker will reference the IoUringSocket. So I think we can keep this order, that may ease some of rebase work when rebase the test PR.

File: test/mocks/io/BUILD
Comment (soulxu) [reply test/mocks/io/BUILD:13]: got it, will remove it.

File: source/common/io/io_uring_worker_impl.h
Comment (KBaichoo) [inline source/common/io/io_uring_worker_impl.h:114]: s/defer to delete./defer deletion./

File: source/common/io/io_uring_worker_impl.cc
Comment (KBaichoo) [inline source/common/io/io_uring_worker_impl.cc:45]: s/existed/existing/

**Review by KBaichoo — COMMENTED** [review]
Sorry for the delay,

Thank you for working on this.

/wait

File: envoy/common/io/io_uring.h
Comment (soulxu) [reply envoy/common/io/io_uring.h:15]: got it, will do

File: envoy/common/io/io_uring.h
Comment (soulxu) [reply envoy/common/io/io_uring.h:193]: Request object is the user data we provide to each iouring request which is used to hold the request related data. Like for read/write, it will carry the buffer, for accept, it will carry the peer address.

https://github.com/envoyproxy/envoy/pull/24958/files#diff-df9a08b7c36ffbd691c64a9c0c22a34cd2848af1ce434bc05bd661dbba4b7199R41

File: source/common/io/io_uring_worker_impl.h
Comment (soulxu) [reply source/common/io/io_uring_worker_impl.h:11]: the Request object is the wrap of request user data, it is part of IoUring API
https://github.com/envoyproxy/envoy/blob/81b11f7bf9d94f557b04749a70240c3fcb40fad1/envoy/common/io/io_uring.h#L95-L96

In the latter PR, we will use it to carry data for each request
https://github.com/envoyproxy/envoy/pull/24958/files#diff-df9a08b7c36ffbd691c64a9c0c22a34cd2848af1ce434bc05bd661dbba4b7199R37-R43

**Comment by soulxu** [comment]
> Sorry for the delay,
> 
> Thank you for working on this.
> 
> /wait

no worries, I'm busy with other stuff this week also, will update the PR and address comments next week. Thanks for the review again!

**Comment by soulxu** [comment]
@zhxie @KBaichoo addressed all the comments, would you like to review it again? thanks!

**Comment by soulxu** [comment]
also note, this passed the tests in https://github.com/envoyproxy/envoy/pull/28824

**Comment by soulxu** [comment]
> also note, this passed the tests in #28824

oh yea, we got all green on all the x86 and arm tests https://github.com/envoyproxy/envoy/pull/28824 now

File: test/common/io/io_uring_impl_test.cc
Comment (KBaichoo) [inline test/common/io/io_uring_impl_test.cc:17]: explicit ctor to avoid implicit conversions

File: test/common/io/io_uring_worker_impl_integration_test.cc
Comment (KBaichoo) [inline test/common/io/io_uring_worker_impl_integration_test.cc:113]: could this be named something more clear such as `createServerListenerAndClientSocket` , it's a bit hard to read otherwise

File: test/common/io/io_uring_worker_impl_integration_test.cc
Comment (KBaichoo) [inline test/common/io/io_uring_worker_impl_integration_test.cc:97]: seems like most of this can be protected visibility

File: test/common/io/io_uring_worker_impl_integration_test.cc
Comment (KBaichoo) [inline test/common/io/io_uring_worker_impl_integration_test.cc:193]: const

File: test/common/io/io_uring_worker_impl_integration_test.cc
Comment (KBaichoo) [inline test/common/io/io_uring_worker_impl_integration_test.cc:220]: seems like the above injects 7 ops, but we only check for 6 here?

**Review by KBaichoo — COMMENTED** [review]
Thank you for working on this.

/wait

File: envoy/common/io/io_uring.h
Comment (soulxu) [reply envoy/common/io/io_uring.h:33]: Yes, this class will be extended. We will have ReadRequest, AcceptRequest etc for carry different data needed by different type request

File: envoy/common/io/io_uring.h
Comment (soulxu) [reply envoy/common/io/io_uring.h:38]: yea, you are right, I forget to change this.

File: source/common/io/io_uring_worker_impl.cc
Comment (soulxu) [reply source/common/io/io_uring_worker_impl.cc:9]: This will be used later when the all the requests done or canceled, then the socket can remove itself from the worker. https://github.com/envoyproxy/envoy/pull/24958/files#diff-a077c8c7da068f59957e43fda10a18909ad921da84175b0d4abdb46da7778990R782

File: source/common/io/io_uring_worker_impl.h
Comment (soulxu) [reply source/common/io/io_uring_worker_impl.h:31]: got it

File: source/common/io/io_uring_worker_impl.h
Comment (soulxu) [reply source/common/io/io_uring_worker_impl.h:56]: Later we will add IoUringServerSocket and IoUringClientSocket, they will log something also.

File: test/common/io/io_uring_impl_test.cc
Comment (soulxu) [reply test/common/io/io_uring_impl_test.cc:17]: got it

File: test/common/io/io_uring_worker_impl_integration_test.cc
Comment (soulxu) [reply test/common/io/io_uring_worker_impl_integration_test.cc:113]: got it, will change it

File: test/common/io/io_uring_worker_impl_integration_test.cc
Comment (soulxu) [reply test/common/io/io_uring_worker_impl_integration_test.cc:220]: emm, good catch, indeed missing one, thanks!

**Review by KBaichoo — APPROVED** [review]
Thanks, lgtm!

**Pull request merged**
